File: C++/代码随想录/Binary_Tree/Find the attributes of the binary search tree/98. Validate Binary Search Tree.cpp
// 98. Validate Binary Search Tree
// https://leetcode.com/problems/validate-binary-search-tree/
// clang++ 98.\ Validate\ Binary\ Search\ Tree.cpp -std=c++17
#include "../../printUtils.hpp"

template<typename T>
class Solution {
public:
    bool isValidBST(TreeNode<T>* root) {
        if (!root) return true;
        T minVal = std::numeric_limits<T>::lowest();
        T maxVal = std::numeric_limits<T>::max();
        return isValidBST(root, minVal, maxVal);
    }

private:
    bool isValidBST(TreeNode<T>* root, T minVal, T maxVal) {
        if (!root) return true;
        
        if (root->val <= minVal || root->val >= maxVal) return false;

        return isValidBST(root->left, minVal, root->val) && isValidBST(root->right, root->val, maxVal);
    }
};

int main() {
    TreeNode<int>* root = buildTree({5, 1, 4, -1, -1, 3, 6}, -1);
    printArray(levelOrder(root));
    std::cout << std::boolalpha << Solution<int>().isValidBST(root) << std::endl;

    std::cout << "====================" << std::endl;
    
    TreeNode<int>* root1 = buildTree({2, 1, 3}, -1);
    printArray(levelOrder(root1));
    std::cout << std::boolalpha << Solution<int>().isValidBST(root1) << std::endl;

    return 0;
}

File: C++/代码随想录/Binary_Tree/Find the attributes of the binary search tree/98. Validate Binary Search Tree.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def validate(node, low=float('-inf'), high=float('inf')):
            if not node:
                return True
            if node.val <= low or node.val >= high:
                return False
            return validate(node.right, node.val, high) and validate(node.left, low, node.val)
        
        return validate(root)

# Test case
root = TreeNode(2, TreeNode(1), TreeNode(3))
sol = Solution()
print(sol.isValidBST(root))  # True
